### ISSUE TITLE
Alignment issues when source cog is in non-native projection

### ISSUE THREAD

**Issue by staffordsmith83** [hongfaqiu/TIFFImageryProvider#16]
We are having problems where COGs are not positioned correctly on the cesium globe when they are reprojected by TIFFImageryProvider:

<img width="629" alt="image" src="https://github.com/hongfaqiu/TIFFImageryProvider/assets/7980991/efc0f926-30a4-4dcc-9e14-e8924b2b0f2f">

This COG is in EPSG:32756, available here: https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/56/J/NP/2023/4/S2A_56JNP_20230410_0_L2A/TCI.tif

**Comment by hongfaqiu** [comment]
I don't know what's going on either, but there may be some issues with the simple bbox's coordinate system conversion. Do you know how to improve it?

**Comment by hongfaqiu** [comment]
Thank you for your help! All the code related to reprojection is here: https://github.com/hongfaqiu/TIFFImageryProvider/blob/main/packages/TIFFImageryProvider/src/TIFFImageryProvider.ts#L200-L227
I have just simply obtained the bounding box of the TIFF, converted it to the bounding box in the 4326 coordinate system, and passed it into the rectangle. May be I should rewrite the entire TilingScheme?

**Comment by staffordsmith83** [comment]
Im not sure, Ill have a look at this tomorrow, perhaps your method is not sufficient. Thanks for your work so far! We are trying to use your code in our open source tool [TerriaJS](https://github.com/TerriaJS/terriajs)

**Comment by hongfaqiu** [comment]
Thank you for your recognition! I really love TerriaJS, and have learned a lot from it.

**Comment by staffordsmith83** [comment]
I have confirmed that reprojecting a cog to EPSG:4326 fixes the alignment problem:
![COG shift TIFFImageryProvider](https://github.com/hongfaqiu/TIFFImageryProvider/assets/7980991/0514f4d2-fbcf-4221-bd67-3c0f265ac3a8)

Datasets
Original cog that appears misaligned: https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/56/J/NP/2023/4/S2A_56JNP_20230410_0_L2A/TCI.tif
Reprojected with gdal and uploaded to S3: https://stafs-cogs.s3.ap-southeast-2.amazonaws.com/Whian_Whian_TCI_4326_cog.tif

**Comment by hongfaqiu** [comment]
Cool! I have found a repo [geotiff-geokeys-to-proj4](https://github.com/matafokka/geotiff-geokeys-to-proj4) that can perform coordinate system transformation on TIFF files. It seems to perform the transformation on every pixel rather than just the bounding box. I'm not sure if this will be effective.

**Comment by staffordsmith83** [comment]
Yes I dont think reprojecting the bounding box is sufficient. This is how OpenLayers handles it: https://openlayers.org/doc/tutorials/raster-reprojection.html

**Comment by hongfaqiu** [comment]
Great news! I solved this problem by creating a custom TilingScheme, you can pass ``projFunc`` to handle it.
```ts
import proj4 from 'proj4-fully-loaded'; 

TIFFImageryProvider.fromUrl(YOUR_TIFF_URL, {
  projFunc: (code) => {
    if (![4326].includes(code)) {
      {
        try {
          let prj = proj4(`EPSG:${code}`, "EPSG:4326")
          let unprj = proj4("EPSG:4326", `EPSG:${code}`)
          if (prj && unprj) return {
            project: prj.forward,
            unproject: unprj.forward
          }
        } catch (e) {
          console.error(e);
        }
      }
    }
  },
});
```
But the demo performance of nextjs was abnormal. I created a vite example and it worked very well. I will solve this problem in the afternoon.

**Comment by staffordsmith83** [comment]
@hongfaqiu this is great news, thanks for your efforts! I will check it out now

**Comment by staffordsmith83** [comment]
Dear @hongfaqiu , I have incorporated your changes into my terriajs test branch, and it resolved the alignment issue, thanks so much for your work! However, it does present a different problem: the COGs are sometimes displayed in the wrong location when zoomed out: 
![shifting position2](https://github.com/hongfaqiu/TIFFImageryProvider/assets/7980991/2311016d-4ea6-43ea-a4fc-cb738a958368)

**Comment by staffordsmith83** [comment]
Dear @hongfaqiu , our cesium expert suggests that you can fix this by changing how `this._rectangle` is computed, to account for the northwest and southwest corners having different longitudes - I think this code should do it:

```
 const swMeters = new Cartesian3();
    options.rectangleSouthwestInMeters.clone(swMeters);
    const neMeters = new Cartesian3();
    options.rectangleNortheastInMeters.clone(neMeters);
    const seMeters = new Cartesian3(neMeters.x, swMeters.y);
    const nwMeters = new Cartesian3(swMeters.x, neMeters.y);

    const southwest = this.projection.unproject(swMeters);
    const southeast = this.projection.unproject(seMeters);
    const northwest = this.projection.unproject(nwMeters);
    const northeast = this.projection.unproject(neMeters);

 this._rectangle = new Rectangle(
      Math.min(
        southwest.longitude,
        southeast.longitude,
        northwest.longitude,
        northeast.longitude
      ),
      Math.min(
        southwest.latitude,
        southeast.latitude,
        northwest.latitude,
        northeast.latitude
      ),
      Math.max(
        southwest.longitude,
        southeast.longitude,
        northwest.longitude,
        northeast.longitude
      ),
      Math.max(
        southwest.latitude,
        southeast.latitude,
        northwest.latitude,
        northeast.latitude
      )
    );
```

**Comment by hongfaqiu** [comment]
Dear @staffordsmith83, thank you and your team for the work you have done. It's working great now! 
There is still a lot of room for improvement in this library, and I look forward to working with you to build it together.

**Comment by hongfaqiu** [comment]
When I project the non-3857 coordinate system into the 4326 coordinate system, I noticed that the four vertices constructed using the bbox attribute of the geotiff do not form a regular rectangle. This also causes the failure of constructing a **Cesium.Rectangle**. 
The TilingScheme only performs projection transformation and mapping on regular rectangular tile boundaries, and it cannot handle tile seams.
I am considering implementing the coordinate transformation of projections in a different way in the frontend. It seems that the TilingScheme approach is incorrect.

**Comment by hongfaqiu** [comment]
I have developed a front-end grid reprojection method, which produces the following results as shown in the image. It addresses some issues but there are still some remaining problems to be resolved.
![动画](https://github.com/hongfaqiu/TIFFImageryProvider/assets/62411296/ad239f11-ed2e-4aa8-b907-a4f87d6ce9e4)

**Comment by staffordsmith83** [comment]
Thanks for your work on this @hongfaqiu , we are following with great interest

**Comment by staffordsmith83** [comment]
Dear @hongfaqiu how did you go resolving these issues? Any luck?

**Comment by hongfaqiu** [comment]
Dear @staffordsmith83, I was busy with other things recently, but today I had some time to update the code. However, the results are not satisfactory, and there are areas for improvement.

Originally, my idea was to make the implementation of the `tileXYToRectangle` function consistent with the `GeographicTilingScheme`, where I would request additional tile data to ensure complete coverage after reprojection. However, this approach encountered some issues, and I had to switch to a solution similar to the `WebMercatorTilingScheme`.

In any case, the reprojection function is currently executing sequentially on the CPU, which makes it inefficient and blocks the rendering process. That's why I have marked the reprojection method as **experimental**.

**Comment by staffordsmith83** [comment]
Dear @hongfaqiu thanks somuch for your efforts! Yes its a difficult issue, the library that we are using for 2d mode, `georaster-layer-for leaflet` also has issues reprojecting. Also busy with other things, but can hopefully come back to this soon. Thanks again!

**Comment by hongfaqiu** [comment]
@staffordsmith83 Thank you very much for your ongoing support. I will also continue to follow your progress in terriaJS and look forward to finding a more effective general solution. It would be even better if this could help advance Cesium's capabilities in reprojection.

**Comment by staffordsmith83** [comment]
Dear @hongfaqiu did you make any progress with the reprojection? I notice in your changelog you have implemented some web based reprojection that is slow but works? Perhaps GeoWarp is an option? https://github.com/DanielJDufour/geowarp

**Comment by hongfaqiu** [comment]
> Dear @hongfaqiu did you make any progress with the reprojection? I notice in your changelog you have implemented some web based reprojection that is slow but works? Perhaps GeoWarp is an option? https://github.com/DanielJDufour/geowarp

Excellent repo! I will try it next week. Thanks your sharing.

**Comment by hongfaqiu** [comment]
I have looked at DanielJDufour's code, and its reprojection is also based on CPU computation. Furthermore, it performs poorly on non-rectangular boundaries as it clips a portion of the content. 
If there is no solution that offers performance improvements of 1-2 orders of magnitude, the feasibility of browser-based reprojection remains low. 
I still have other work tasks this week, so I hope to have some free time next week to investigate this issue.
